requirements.txt

```diff
@@ -1,4 +1,6 @@
 ops >= 1.2.0
 kubernetes
 jinja2
-pyyaml
\ No newline at end of file
+pyyaml
+git+https://opendev.org/openstack/charm-ops-interface-ceph-client@d4d0cefc5e92edd28cdbf5d00e190d4498896418#egg=interface_ceph_client
+netifaces
```

Comment:
What's this for?  I can't find it used in this PR.  Does the above requirement need it but not list it in the `setup.py`

Reply:
oh, I see it's used in the [`test-requirements.txt`](https://opendev.org/openstack/charm-ops-interface-ceph-client/src/branch/master/test-requirements.txt#L7) of that upstream library.  It's clear i don't quite grok it's necessity but i'll +1

Reply:
Oh yeah! Let me show you what happens when this is missing:

```
$ tox -e unit
...
Reading package lists... Done
E: Could not open lock file /var/lib/apt/lists/lock - open (13: Permission denied)
E: Unable to lock directory /var/lib/apt/lists/
W: Problem unlinking the file /var/cache/apt/pkgcache.bin - RemoveCaches (13: Permission denied)
W: Problem unlinking the file /var/cache/apt/srcpkgcache.bin - RemoveCaches (13: Permission denied)
juju-log: Couldn't acquire DPKG lock. Will retry in 10 seconds
^CE..........................
======================================================================
ERROR: test_charm (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: test_charm
Traceback (most recent call last):
  File "/home/gkk/work/tasks/bug-ceph-csi/ceph-csi-operator/.tox/unit/lib/python3.8/site-packages/charmhelpers/contrib/network/ip.py", line 38, in <module>
    import netifaces
ModuleNotFoundError: No module named 'netifaces'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/gkk/work/tasks/bug-ceph-csi/ceph-csi-operator/.tox/unit/lib/python3.8/site-packages/charmhelpers/fetch/ubuntu.py", line 922, in _run_with_retries
    result = subprocess.check_call(cmd, env=env, **kwargs)
  File "/usr/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['apt-get', 'update']' returned non-zero exit status 100.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.8/unittest/loader.py", line 436, in _find_test_path
    module = self._get_module_from_name(name)
  File "/usr/lib/python3.8/unittest/loader.py", line 377, in _get_module_from_name
    __import__(name)
  File "/home/gkk/work/tasks/bug-ceph-csi/ceph-csi-operator/tests/unit/test_charm.py", line 27, in <module>
    from interface_ceph_client.ceph_client import CephClientRequires
  File "/home/gkk/work/tasks/bug-ceph-csi/ceph-csi-operator/.tox/unit/lib/python3.8/site-packages/interface_ceph_client/ceph_client.py", line 8, in <module>
    import charmhelpers.contrib.storage.linux.ceph as ch_ceph
  File "/home/gkk/work/tasks/bug-ceph-csi/ceph-csi-operator/.tox/unit/lib/python3.8/site-packages/charmhelpers/contrib/storage/linux/ceph.py", line 69, in <module>
    from charmhelpers.contrib.openstack.utils import config_flags_parser
  File "/home/gkk/work/tasks/bug-ceph-csi/ceph-csi-operator/.tox/unit/lib/python3.8/site-packages/charmhelpers/contrib/openstack/utils.py", line 34, in <module>
    from charmhelpers.contrib.network import ip
  File "/home/gkk/work/tasks/bug-ceph-csi/ceph-csi-operator/.tox/unit/lib/python3.8/site-packages/charmhelpers/contrib/network/ip.py", line 40, in <module>
    apt_update(fatal=True)
  File "/home/gkk/work/tasks/bug-ceph-csi/ceph-csi-operator/.tox/unit/lib/python3.8/site-packages/charmhelpers/fetch/ubuntu.py", line 400, in apt_update
    _run_apt_command(cmd, fatal)
  File "/home/gkk/work/tasks/bug-ceph-csi/ceph-csi-operator/.tox/unit/lib/python3.8/site-packages/charmhelpers/fetch/ubuntu.py", line 945, in _run_apt_command
    _run_with_retries(
  File "/home/gkk/work/tasks/bug-ceph-csi/ceph-csi-operator/.tox/unit/lib/python3.8/site-packages/charmhelpers/fetch/ubuntu.py", line 929, in _run_with_retries
    time.sleep(CMD_RETRY_DELAY)
KeyboardInterrupt
```

Bwuh? An apt command is running, and before the tests even started? Yep, importing `interface_ceph_client` takes us on a charmhelpers import trail that eventually leads to apt calls [here](https://github.com/juju/charm-helpers/blob/a72931f7324526e7f05221847437238517f38fa7/charmhelpers/contrib/network/ip.py#L37-L42).

I don't want that happening in the tests *or* at run time, so I put it in requirements.txt. Hope that makes sense.

Reply:
FUN!  glad you caught that one

Reply:
I added a comment to hopefully clarify this for future codebase readers, please take a look.